backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/ReflectionSupport.kt

```diff
@@ -10,95 +10,160 @@ import org.jetbrains.kotlin.backend.konan.KonanBackendContext
 import org.jetbrains.kotlin.ir.symbols.IrClassSymbol
+import org.jetbrains.kotlin.ir.symbols.IrTypeParameterSymbol
 import org.jetbrains.kotlin.ir.types.*
-import org.jetbrains.kotlin.ir.util.constructors
+import org.jetbrains.kotlin.ir.util.*
 import org.jetbrains.kotlin.ir.util.irCall
 import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameUnsafe
 import org.jetbrains.kotlin.resolve.descriptorUtil.getAllSuperClassifiers
+import org.jetbrains.kotlin.types.Variance
 
 internal class KTypeGenerator(
-        private val context: KonanBackendContext,
-        private val eraseTypeParameters: Boolean = false
+        val context: KonanBackendContext,
+        val irFile: IrFile,
+        val irElement: IrElement,
+        val needExactTypeParameters: Boolean = false
 ) {
     private val symbols = context.ir.symbols
 
-    fun IrBuilderWithScope.irKType(type: IrType): IrExpression = if (type !is IrSimpleType) {
-        // Represent as non-denotable type:
-        irKTypeImpl(
-                kClassifier = irNull(),
-                irTypeArguments = emptyList(),
-                isMarkedNullable = false
-        )
-    } else {
-        val classifier = type.classifier
+    fun IrBuilderWithScope.irKType(type: IrType, leaveReifiedForLater: Boolean = false) =
+            irKType(type, BuildingContext(leaveReifiedForLater, mutableSetOf()))
 
-        if (classifier is IrClassSymbol) {
-            irKTypeImpl(
-                    kClassifier = irKClass(classifier),
-                    irTypeArguments = type.arguments,
-                    isMarkedNullable = type.hasQuestionMark
+    private class BuildingContext(var leaveReifiedForLater: Boolean, val seenTypeParameters: MutableSet<IrTypeParameter>)
+
+    private class RecursiveBoundsException(message: String) : Throwable(message)
+
+    private fun IrBuilderWithScope.irKType(type: IrType, context: BuildingContext): IrExpression {
+        if (type !is IrSimpleType) {
+            // Represent as non-denotable type:
+            return irKTypeImpl(
+                    kClassifier = irNull(),
+                    irTypeArguments = emptyList(),
+                    isMarkedNullable = false,
+                    context = context
             )
-        } else {
-            if (eraseTypeParameters) {
-                irKType(context.irBuiltIns.anyNType)
-            } else {
-                irCall(symbols.kTypeImplForGenerics.constructors.single())
+        }
+        try {
+            val kClassifier = when (val classifier = type.classifier) {
+                is IrClassSymbol -> irKClass(classifier)
+                is IrTypeParameterSymbol -> {
+                    if (classifier.owner.isReified && context.leaveReifiedForLater) {
+                        // Leave as is for reification.
+                        return irCall(symbols.typeOf).apply { putTypeArgument(0, type) }
+                    }
+
+                    context.leaveReifiedForLater = false
```

Comment:
What is this required for?

Reply:
To leave upper bounds of non-reified type parameters as is, even if they are reified themselves (as JVM does).

Reply:
I see.
Shouldn't we revert this after `irKTypeParameter` below then?

Reply:
Indeed. Previously it was just a parameter. I guess I'll revert this since now `BuildingContext` contains only two values.